server/src/main/java/org/opensearch/common/cache/Cache.java

```diff
@@ -782,10 +810,21 @@ private boolean promote(Entry<K, V> entry, long now) {
                     break;
             }
             if (promoted) {
-                evict(now);
+                while (tail != null && shouldPrune(tail, now)) {
+                    Entry<K, V> entryToBeRemoved = tail;
+                    CacheSegment<K, V> segment = getCacheSegment(entryToBeRemoved.key);
+                    if (segment != null) {
+                        segment.remove(entryToBeRemoved.key, entryToBeRemoved.value, f -> {});
+                    }
+                    if (unlink(entryToBeRemoved)) {
+                        removalNotifications.add(
+                            new RemovalNotification<>(entryToBeRemoved.key, entryToBeRemoved.value, RemovalReason.EVICTED)
+                        );
+                    }
```

Comment:
So, we still remove the entries under the global lock, but don't send the notifications until after the lock has been released. What does this do to tests for cache statistics? In practice, probably nothing. When a given call to `get` or `put` completes, the notifications have been processed -- they're not 

Also, I think the old behavior meant that a `removalListener` associated with a single cache would definitely have `onRemoval` called by a single thread. Of course, that thread-safety wasn't documented anywhere. (The `RemovalListener` interface has no Javadoc, so 🤷 .) I think I managed to find all of the `RemovalListener`s related to `Cache`s (thankfully they all get wired in via `CacheBuilder`).  From what I've seen, they're all thread-safe, doing atomic increments/decrements (which makes sense, since they could be read at any time).

Can you please add a comment to `RemovalListener#onRemoval` to make it clear that `onRemoval` may be called from multiple threads at once, and therefore implementers should be thread-safe?